xarray/backends/netCDF4_.py

```diff
@@ -25,14 +26,29 @@
                   '|': 'native'}
 
 
+class VariableWrapper(object):
+    """ Ensures that variable attributes and data access operates on open datastore """
+    def __init__(self, datastore, variable_name):
+        self.var = variable_name
+        self.datastore = datastore
+
+    def __getattr__(self, attr):
+        with self.datastore.ensure_open():
+            return getattr(self.datastore.ds.variables[self.var], attr)
+
+    def __getitem__(self, key):
+        with self.datastore.ensure_open():
+            return operator.getitem(self.datastore.ds.variables[self.var], key)
+
+
 class BaseNetCDF4Array(NDArrayMixin):
     def __init__(self, variable_name, datastore):
         self.datastore = datastore
         self.variable_name = variable_name
 
     @property
     def array(self):
-        return self.datastore.ds.variables[self.variable_name]
+        return VariableWrapper(self.datastore, self.variable_name)
```

Comment:
@shoyer, this is confusing to me because there are times when the `array` property is needed for a variable.  If the dataset is closed, this causes an error (hence why this code is here).  Are you suggesting we should basically cache `array.shape` and `array.dtype`?  I think I'm missing something more important here, something like `shape` and `dtype` will be read from the base array `BaseNetCDF4Array` without need for `array` if they exist.

I fully agree that there is a lot of churn when handling a large number of files, but it isn't obvious to me how to avoid this issue.